File: internal/cred/claude.go
package cred

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
)

const ClaudeTokenMissingMessage = "claude token not found — run `claude /login` to authenticate"

var ErrClaudeTokenNotFound = errors.New(ClaudeTokenMissingMessage)

// ErrClaudeWriteUnsupported is returned by WriteClaudeLiveBlob on platforms
// that do not support writing the Claude live credential (anything other than
// macOS or Linux).
var ErrClaudeWriteUnsupported = errors.New("writing Claude live credential is not supported on this platform")

// Credential holds the parsed fields of a provider's live credential blob plus
// the exact input bytes. Raw preserves the original payload byte-for-byte so
// callers can pass it onward (e.g. to WriteClaudeLiveBlob or accounts.Account)
// without re-marshaling and without risking field loss.
//
// Immutability convention: Raw is set once by parseClaudeCredFull via
// bytes.Clone and must not be mutated afterward. Callers that need to pass Raw
// to a writer should treat it as read-only.
type Credential struct {
	AccessToken  string
	RefreshToken string
	// ExpiresAt is ms since epoch; 0 if absent. Claude only — populated from the
	// claudeAiOauth.expiresAt field. Codex leaves this 0 (its auth.json has no
	// expiry field); the codex refresh gate decodes the access-token JWT exp on
	// demand in codex.StoredExpiresAt.
	ExpiresAt int64
	Raw       []byte
}

// parseClaudeCredFull parses the JSON payload used by both the macOS Keychain
// item ("Claude Code-credentials") and the Linux ~/.claude/.credentials.json
// file. It validates that the access token is present and copies the input
// bytes verbatim into Credential.Raw via bytes.Clone so subsequent mutations
// of the caller's buffer cannot corrupt the stored copy.
//
// Returns ErrClaudeTokenNotFound when the access token field is empty or
// absent. RefreshToken and ExpiresAt are optional; they are zero on absence.
func parseClaudeCredFull(data []byte) (Credential, error) {
	var raw struct {
		ClaudeAiOauth struct {
			AccessToken  string `json:"accessToken"`
			RefreshToken string `json:"refreshToken"`
			ExpiresAt    int64  `json:"expiresAt"`
		} `json:"claudeAiOauth"`
	}
	if err := json.Unmarshal(data, &raw); err != nil {
		return Credential{}, fmt.Errorf("claude credential is not valid JSON: %w", err)
	}
	if raw.ClaudeAiOauth.AccessToken == "" {
		return Credential{}, ErrClaudeTokenNotFound
	}
	return Credential{
		AccessToken:  raw.ClaudeAiOauth.AccessToken,
		RefreshToken: raw.ClaudeAiOauth.RefreshToken,
		ExpiresAt:    raw.ClaudeAiOauth.ExpiresAt,
		Raw:          bytes.Clone(data),
	}, nil
}

// parseClaudeCred extracts the OAuth access token from the JSON payload used
// by both the macOS Keychain item ("Claude Code-credentials") and the Linux
// ~/.claude/.credentials.json file. Returns ErrClaudeTokenNotFound when the
// token field is empty or absent.
func parseClaudeCred(data []byte) (string, error) {
	c, err := parseClaudeCredFull(data)
	return c.AccessToken, err
}

File: src/cred/claude.rs
use super::{Credential, CredError};
use serde::Deserialize;

#[derive(Deserialize)]
struct ClaudeCred {
    #[serde(rename = "claudeAiOauth")]
    claude_ai_oauth: ClaudeOAuth,
}

#[derive(Deserialize)]
struct ClaudeOAuth {
    #[serde(rename = "accessToken")]
    access_token: Option<String>,
    #[serde(rename = "refreshToken")]
    refresh_token: Option<String>,
    #[serde(rename = "expiresAt")]
    expires_at: Option<i64>,
}

pub fn parse_claude_cred(data: &[u8]) -> Result<Credential, CredError> {
    let raw: ClaudeCred = serde_json::from_slice(data)
        .map_err(|e| CredError::Other(format!("claude credential is not valid JSON: {e}")))?;
    let access_token = raw
        .claude_ai_oauth
        .access_token
        .filter(|t| !t.is_empty())
        .ok_or(CredError::ClaudeNotFound)?;
    Ok(Credential {
        access_token,
        refresh_token: raw.claude_ai_oauth.refresh_token.unwrap_or_default(),
        expires_at: raw.claude_ai_oauth.expires_at.unwrap_or(0),
        raw: data.to_vec(),
    })
}

pub fn read_claude_credential() -> Result<Credential, CredError> {
    #[cfg(target_os = "macos")]
    {
        super::keychain_darwin::read_claude_credential()
    }
    #[cfg(target_os = "linux")]
    {
        super::keychain_linux::read_claude_credential()
    }
    #[cfg(not(any(target_os = "macos", target_os = "linux")))]
    {
        Err(CredError::Other(
            "reading Claude credential not supported on this platform".into(),
        ))
    }
}

pub fn write_claude_live_blob(raw_blob: &[u8]) -> Result<(), CredError> {
    #[cfg(target_os = "macos")]
    {
        super::keychain_darwin::write_claude_live_blob(raw_blob)
    }
    #[cfg(target_os = "linux")]
    {
        super::keychain_linux::write_claude_live_blob(raw_blob)
    }
    #[cfg(not(any(target_os = "macos", target_os = "linux")))]
    {
        let _ = raw_blob;
        Err(CredError::Other(
            "writing Claude live credential is not supported on this platform".into(),
        ))
    }
}
